### ISSUE TITLE
Google login - InvalidStateException

### ISSUE THREAD

**Issue by vk011** [laravel/socialite#389]
- Socialite Version: "laravel/socialite": "^4.1",
- Laravel Version: 5.8
- PHP Version: 7. 2.18
- Database Driver & Version:

### Description:
Specifically for Google (Facebook works) I get an `InvalidStateException` .

However if I comment out ``profile` under `/vendor/laravel/socialite/src/Two/GoogleProvider.php`

so it looks like:

```
protected $scopes = [
        'openid',
        //'profile',
        'email',
    ];
```

Then login works, however then I don't have access to `name` which I really need.

**Comment by theodh** [comment]
@vk011 couldn't reproduce this with my environment:
PHP: 7.2.20
Laravel: v5.8.35
Socialite: 4.1.4

Scope profile is working.

Perhaps you used an old token, I have found this [forum post](https://stackoverflow.com/questions/20555351/google-oauth2-re-authorization-is-missing-permissions-on-the-consent-page)

**Comment by vk011** [comment]
Hi,

I don't believe I am using an older token as this happens whenever I check and has been happening from the very beginning, though I do not really know how I would check which token is being used (whether old or not).

If I add `stateless()` in `handleProviderCallback` like so:
`$user = Socialite::driver($provider)->stateless()->user();`

then I get the following error, does that maybe tell you anything useful?

```
Client error: `POST https://www.googleapis.com/oauth2/v4/token` resulted in a `400 Bad Request` response: { "error": "invalid_request", "error_description": "Missing required parameter: code" }
```

**Comment by theodh** [comment]
Hi @vk011 

Tried again, couldn't reproduce it. 

For your handleProviderCallback you should get the following request back, before you'r post the token:
```
https://www.example.com/your_handle_request_google?state=your_state&code=code_from_google&scope=email+profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile+openid+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.email&authuser=1&session_state=session_key&prompt=none#
```

Is there a param code in here, first line of your handleProviderCallback:
```
dd(Input::get('code'));
```

**Comment by driesvints** [comment]
Hey there,

Can you first please try one of the support channels below? If you can actually identify this as a bug, feel free to report back and I'll gladly help you out and re-open this issue.

- [Laracasts Forums](https://laracasts.com/discuss)
- [Laravel.io Forums](https://laravel.io/forum)
- [StackOverflow](https://stackoverflow.com/questions/tagged/laravel)
- [Discord](https://discordapp.com/invite/KxwQuKb)
- [Larachat](https://larachat.co)
- [IRC](https://webchat.freenode.net/?nick=laravelnewbie&channels=%23laravel&prompt=1)

Thanks!

**Comment by mmjjb** [comment]
@driesvints I have the same issue. I just did an fresh installation of Socialite and ran straight into `InvalidStateException`; if I add `stateless()` I receive `400 Bad Request` on `POST https://www.googleapis.com/oauth2/v4/token`

**Comment by vk011** [comment]
@mmjjb And if you comment out `profile` it works right?

**Comment by vk011** [comment]
@theodh It does even though it says it doesn't, but anyway that's when using `stateless()` which shouldn't be used in this case anyway, anyhow this is the query string of the failed login attempt (without using stateless)

https://mysite.com/login/google/callback?
state=tgrqoFXksk8j32dzIKe2gzLh5V5Dcfvrwo0hYP7F&code=4%2FsAH8v9lPjiBHM0RmsvZbE398P8jVEaqEl5HrOY8mAtQC2rKbbjyWPtA2rt1FRtAuRbm2L6en5_OF0WwAOZXy9aQ&
scope=email+profile+openid+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.email&authuser=0&session_state=62ea6651720b6e52a2dd671aedee2b81f5c2925b..18ad&prompt=consent#

**Comment by jomarierafa** [comment]
it also a problem on me on facebook login, i also add stateless and it produce new error 
POST https://graph.facebook.com/v3.0/oauth/access_token` resulted in a `400 Bad Request` response: {"error":{"message":"Missing authorization code"

**Comment by Softelicious** [comment]
For me it's also happening, idk how it works, but if I change driver() to with() it starts working for a day, then it stops, then I change session table to array and it starts working, and stops later on, now I added stateless() after driver() and its started to work again, idk for how long

**Comment by vk011** [comment]
@xf039 Are you by any change using an addon domain for that specific website?

**Comment by niladam** [comment]
I'm also affected by this. However, i think there's an issue with the permissions the user is accepting/denying. I tried on my own account and it works flawlessly. I'm posting this here to make sure that i'm being notified when others find a solution.

**Comment by thomas-cruz** [comment]
I have the same issue. I can add stateless() to get around it, but now it results in 400 bad request missing required parameter: code

**Comment by vk011** [comment]
@thomas-cruz Are you using an addon domain?

**Comment by thomas-cruz** [comment]
I'm just running it locally on my machine.

**Comment by dilpreet91** [comment]
@vk011  Did you get solution for your problem?

**Comment by vk011** [comment]
@dilpreet91 No, I guess profile just won't work on addon domains for some reason.

**Comment by dilpreet91** [comment]
I just tried with `stateless()` 
`public function redirectToGoogle(){
        return Socialite::driver('google')->stateless()->redirect();
    }
 public function googleLogin(Request $request)
    {
        try {
                $user = Socialite::driver('google')->stateless()->user();
        } catch (\Exception $e) {
          dd($e);
        }
`
I got the user detail along with profile data

**Comment by workzentre** [comment]
Same here... It works well on localhost, without SESSION_DOMAIN, without stateless(), and any other possible suggestion I found. but Live in production, I end up with the same Laravel\Socialite\Two\InvalidStateException
vendor/laravel/socialite/src/Two/AbstractProvider.php:210

**Comment by liepumartins** [comment]
Does this has anything to do with 'same_site' setting for session cookies, as mentioned on [this SO answer](https://stackoverflow.com/a/60997974)?

**Comment by cre-mer** [comment]
I tried @dilpreet91 solution and it worked locally. You have to use stateless twice. Once while redirecting to google and once when you're redirecting back to your website.

**Comment by vk011** [comment]
@cre-mer Are you using an addon domain for that specific website?

**Comment by Mobashir-Monim** [comment]
Whenever I try with the state-full way, it works on localhost but not on the server.
I had to use in stateless for the deployment.

Is there a particular reason why including state creates an error only in production? Or am I missing something?

**Comment by stephan-v** [comment]
The documentation does not really go in depth about `stateless()` but could someone please explain me why this is needed for the Google socialite driver. It only works when I add it both to the redirect and handle.

**Comment by Thommis** [comment]
> Hi,
> 
> I don't believe I am using an older token as this happens whenever I check and has been happening from the very beginning, though I do not really know how I would check which token is being used (whether old or not).
> 
> If I add `stateless()` in `handleProviderCallback` like so: `$user = Socialite::driver($provider)->stateless()->user();`
> 
> then I get the following error, does that maybe tell you anything useful?
> 
> ```
> Client error: `POST https://www.googleapis.com/oauth2/v4/token` resulted in a `400 Bad Request` response: { "error": "invalid_request", "error_description": "Missing required parameter: code" }
> ```

I have the same issue here with the Microsoft Azure driver ... did you get any solution to this issue?

**Comment by aralvi** [comment]
i have the same issue on live server but it works absolutely fine on local server

**Comment by CHamalainen** [comment]
For anyone who might encounter this later, I had a similar problem, and after a very thorough and frustrating investigation I found a solution. I wanna leave it here in case it helps some of you, who might find it later! 

For me the `same_site` cookie setting in the config/session.php file was set to "strict" for my project (the possible options being "lax", "strict", "none"). **Changing the setting to "lax" solved the issue in my case.**

The breakdown of why it might help:
"strict": no cookie can be set from outside the cookie domain
"lax": cookies can be set from outside the cookie domain if method is GET/HEAD and navigation is "Top Level"
"none": Wild wild west, everything is permitted, nothing is certain

https://stackoverflow.com/questions/59990864/what-is-the-difference-between-samesite-lax-and-samesite-strict

**Comment by yusufhikam** [comment]
> Hi,
> 
> I don't believe I am using an older token as this happens whenever I check and has been happening from the very beginning, though I do not really know how I would check which token is being used (whether old or not).
> 
> If I add `stateless()` in `handleProviderCallback` like so: `$user = Socialite::driver($provider)->stateless()->user();`
> 
> then I get the following error, does that maybe tell you anything useful?
> 
> ```
> Client error: `POST https://www.googleapis.com/oauth2/v4/token` resulted in a `400 Bad Request` response: { "error": "invalid_request", "error_description": "Missing required parameter: code" }
> ```

It may be a little late in 2026, but hopefully this will help.

**I got the same error after selecting my Google account :**

> `The GET method is not supported for route auth/login. Supported methods: POST.`

> http://localhost:8000/auth/login?error=Client+error%3A+%60POST+https%3A%2F%2Fwww.googleapis.com%2Foauth2%2Fv4%2Ftoken%65+resulted+in+a+%60400+Bad+Request%60+response%3A%0A%7B%0A++%22error%22%3A+%22invalid_request%22%2C%0A++%22error_description%22%3A+%22Missing+required+parameter%3A+code%22%0A%7D%0A



so I try to running commands:

`php artisan route:clear`
`php artisan cache:clear`
`php artisan config:clear`


After doing that, it worked for me.
